File: player/ultraPlayer/EchoSoundFx.cpp
#include "EchoSoundFx.h"

EchoSoundFx::EchoSoundFx(void):ASoundFx(SoundFxType::Echo)
{
	_wetDryMix = 0;
	_feedback = 0;
	_leftDelay = 0;
	_rightDelay = 0;
	_panDelay = 0;
}

EchoSoundFx::~EchoSoundFx(void)
{
}

float EchoSoundFx::WetDryMix()
{
	return _wetDryMix;
}

void EchoSoundFx::WetDryMix( float val )
{
	_wetDryMix = val;
}

float EchoSoundFx::Feedback()
{
	return _feedback;
}

void EchoSoundFx::Feedback( float val )
{
	_feedback = val;
}

float EchoSoundFx::LeftDelay()
{
	return _leftDelay;
}

void EchoSoundFx::LeftDelay( float val )
{
	_leftDelay = val;
}

float EchoSoundFx::RightDelay()
{
	return _rightDelay;
}

void EchoSoundFx::RightDelay( float val )
{
	_rightDelay = val;
}

float EchoSoundFx::PanDelay()
{
	return _panDelay;
}

void EchoSoundFx::PanDelay( float val )
{
	_panDelay = val;
}

void EchoSoundFx::FromString( string data )
{
	try
	{
		int paramLength = CommunicationProtocol::Instance()->LongNumericParameterLength();
		int index = CommunicationProtocol::Instance()->HeaderLength();

		_wetDryMix = NumericToStringConverter::Convert<float>(data.substr(index, paramLength));
		index += paramLength;
		_feedback = NumericToStringConverter::Convert<float>(data.substr(index, paramLength));
		index += paramLength;
		_leftDelay = NumericToStringConverter::Convert<float>(data.substr(index, paramLength));
		index += paramLength;
		_rightDelay = NumericToStringConverter::Convert<float>(data.substr(index, paramLength));
		index += paramLength;
		_panDelay = NumericToStringConverter::Convert<float>(data.substr(index, paramLength));
	}
	catch (exception& e)
	{

	}
}

std::string EchoSoundFx::ToString()
{
	throw NotImplementedException("","EchoSoundFx::ToString");
}

ASoundFx* EchoSoundFx::MakeCopy()
{
	return (ASoundFx*)(new EchoSoundFx(*this));
}

void EchoSoundFx::CopyFrom( ASoundFx* original )
{
	_wetDryMix = ((EchoSoundFx*)original)->WetDryMix();
	_feedback = ((EchoSoundFx*)original)->Feedback();
	_leftDelay = ((EchoSoundFx*)original)->LeftDelay();
	_rightDelay = ((EchoSoundFx*)original)->RightDelay();
	_panDelay = ((EchoSoundFx*)original)->PanDelay();

}

File: controller/Model/SoundFx/EchoSoundFx.cs
using System;
using System.Collections.Generic;
using System.Text;
using UltraPlayerController.Resources.Communication;
using Utilities.Math;
using UltraPlayerController.Model.Utilities;
using Utilities.Serialize;

namespace UltraPlayerController.Model.SoundFx
{
    public class EchoSoundFx : ASoundFx
    {
        #region Fields
        private MathValue _wetDryMix;
        private MathValue _feedback;
        private MathValue _leftDelay;
        private MathValue _rightDelay;
        private MathValue _panDelay;
        #endregion

        #region Constructors
        public EchoSoundFx()
            : base(SoundFxType.Echo)
        {
            _wetDryMix = new MathValue();
         _feedback = new MathValue();
         _leftDelay = new MathValue();
         _rightDelay = new MathValue();
         _panDelay = new MathValue();

         _wetDryMix.Min = 0; // DSFXECHO_WETDRYMIX_MIN
         _wetDryMix.Max = 100; // DSFXECHO_WETDRYMIX_MAX
            _wetDryMix.Default = 50; 
            _wetDryMix.Unit = MathUnitType.percent ;
            _wetDryMix.FriendlyName = "WetDryMix";
            _wetDryMix.Step = 0.1f;

            _feedback.Min = 0; // DSFXECHO_FEEDBACK_MIN
            _feedback.Max = 100; // DSFXECHO_FEEDBACK_MAX
            _feedback.Default = 50; 
            _feedback.Unit = MathUnitType.percent ;
            _feedback.FriendlyName = "Feedback";
            _feedback.Step = 0.1f;

            _leftDelay.Min = 1; // DSFXECHO_LEFTDELAY_MIN
            _leftDelay.Max = 2000; // DSFXECHO_LEFTDELAY_MAX
            _leftDelay.Default = 500; 
            _leftDelay.Unit = MathUnitType.ms ;
            _leftDelay.FriendlyName = "LeftDelay";
            _leftDelay.Step = 0.1f;

            _rightDelay.Min = 1; // DSFXECHO_RIGHTDELAY_MIN
            _rightDelay.Max = 2000; // DSFXECHO_RIGHTDELAY_MAX
            _rightDelay.Default = 500; 
            _rightDelay.Unit = MathUnitType.ms ;
            _rightDelay.FriendlyName = "RightDelay";
            _rightDelay.Step = 0.1f;

            _panDelay.Min = 0; // DSFXECHO_PANDELAY_MIN
            _panDelay.Max = 1; // DSFXECHO_PANDELAY_MAX
            _panDelay.Default = 0; 
            _panDelay.Unit = MathUnitType.boolean ;
            _panDelay.FriendlyName = "PanDelay";
            _panDelay.Step = 0.1f;

        }
        #endregion

        #region Properties

        public MathValue WetDryMix
        {
            get { return _wetDryMix; }
            set { _wetDryMix = value; }
        }
        public MathValue Feedback
        {
            get { return _feedback; }
            set { _feedback = value; }
        }
        public MathValue LeftDelay
        {
            get { return _leftDelay; }
            set { _leftDelay = value; }
        }
        public MathValue RightDelay
        {
            get { return _rightDelay; }
            set { _rightDelay = value; }
        }
        public MathValue PanDelay
        {
            get { return _panDelay; }
            set { _panDelay = value; }
        }
        #endregion

        #region ASoundFx Members

        public override string ParseToString()
        {
            string result = string.Empty;

            try
            {
                result += base.ParseToString();

                result += ValueSerializer.DoFloat(Protocol.Instance.LongNumericParameterLength, _wetDryMix.Precision(), _wetDryMix.Value);
                result += ValueSerializer.DoFloat(Protocol.Instance.LongNumericParameterLength, _feedback.Precision(), _feedback.Value);
                result += ValueSerializer.DoFloat(Protocol.Instance.LongNumericParameterLength, _leftDelay.Precision(), _leftDelay.Value);
                result += ValueSerializer.DoFloat(Protocol.Instance.LongNumericParameterLength, _rightDelay.Precision(), _rightDelay.Value);
                result += ValueSerializer.DoFloat(Protocol.Instance.LongNumericParameterLength, _panDelay.Precision(), _panDelay.Value);

            }
            catch (System.Exception)
            {
                // throw
            }

            return result;
        }

        public override void ParseFromString(string textToParse)
        {
            throw new NotImplementedException();
        }

        #endregion
    }
}
